soc: nrf54h20: fix some power consumption issues by managing GD domains on certain scenarios

gmarull opened the pull request:
This PR adds a new API to request/release GD domains, so that some identified issues can be handled.

WIP, subject to change. This also requires a new nrf-regtool release.

gmarull committed:
dts: bindings: power: add nordic,nrf-global-pd

Add binding for Global Power Domain found in nRF54Hx SoCs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: add power domain information

So that it can be used to manually control certain power domains.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf54h: add SoC level API to request/release GPD

Add a new soc-level API that allows to manually request/release global
power domains.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pinctrl: nrf: add flag to signal the FAST_ACTIVE1 peripherals

This patch introduces a new flag to indicate if a peripheral belongs
to FAST_ACTIVE1 domain. This way, pinctrl knows when to request the
SLOW_ACTIVE domain (where CTRLSEL multiplexer resides).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: gpio: nrfx: add support for ports with pad on FAST_ACTIVE1

This patch allows to _safely_ configure GPIO ports that have their pad
on FAST_ACTIVE1 domain.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: nordic: nrf54h20: define global power domain

Add the global power domain entry. This domain is not memory-mapped
but controlled using NRFS services.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

karstenkoenig commented on soc/nordic/nrf54h/gpd/gpd.c:
```diff
@@ -0,0 +1,229 @@
+	/* flag GPD service as ready */
+	atomic_set_bit(&gpd_service_status, GPD_SERVICE_READY);
+
+	/* release any domain that was not left active during the boot process */
```
The double reference counting between "before GPD_SERVICE_READY and after GPD_SERVICE_READY" isn't correct, you'd need to inject the different reference counters such as `fast_active1_cnt` into the respective onoff manager by recalling `nrf_gpd_request()` multiple times to make sure all the positive requests are still there, otherwise when something turns off again later that was turned on but not tracked it might lead to an invalid off transition.

Or drop the double reference counting and just rely on onoff manager. In the start/stop transitions you can check if the nrfs backend is available. That means you'd need two init stages for the gpd, one early to initialize the onoff managers and one late to send the state after nrfs is connected and send any state that wasn't transmitted yet

hubertmis commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -95,6 +106,18 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+			return ret;
+		}
+
+		retain = nrf_gpio_port_retain_get(cfg->port);
+		nrf_gpio_port_retain_set(cfg->port, retain | BIT(pin));
```
This seems to be applied too early. When you apply retention, the following operations like `nrf_gpio_port_out_` or `nrf_gpio_reconfigure()` have no effect on the pins.

The retention should be applied after the pin is fully configured. Right after applying the retention, `nrf_gpd_release()` should be called to allow for low-power states.

hubertmis commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -171,7 +197,18 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+end:
+#ifdef CONFIG_SOC_NRF54H20_GPD
+	if (cfg->pad_pd == NRF_GPD_FAST_ACTIVE1) {
+		nrf_gpio_port_retain_set(cfg->port, retain);
```
`| BIT(pin)` seems missing, so the retention won't be applied to the pin we would like to apply

hubertmis reviewed (changes requested):
I think we also need something in pinctrl similar to what is implemented in gpio_nrfx.c. We need to safely control fast pins when fast peripherals get deactivated. It requires GPIO RETAIN feature

hubertmis commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -171,7 +191,21 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+	if (cfg->pad_pd == NRF_GPD_FAST_ACTIVE1) {
+		uint32_t retain;
+
+		retain = nrf_gpio_port_retain_get(cfg->port) | BIT(pin);
```
Yes, the retention should be cleared between `ret = nrf_gpd_request(NRF_GPD_FAST_ACTIVE1);` gets applied and the first attempt to reconfigure pins (`if (flags & GPIO_OUTPUT_INIT_HIGH) { nrf_gpio_port_out_set(cfg->port, BIT(pin));`).

If pins are reconfigured in any other function in this module, it should handle retention in a similar way.

hubertmis commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -367,6 +370,24 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+						return ret;
+					}
+				} else {
+					ret = nrf_gpd_request(NRF_GPD_SLOW_ACTIVE);
```
Now I'm thinking if we should do the GPIO.RETAINSET trick here instead. If we keep GDP requested, it would prevent entering low-power modes

hubertmis replied:
Perhaps not instead, but alongside. We should check if the pin is going to be controlled by the peripheral. If so, request GPD. If not, set GPIO.RETAIN.
And then we would need to identify where to clear GPIO.RETAIN.

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -820,6 +855,7 @@
 				compatible = "nordic,nrf-dppic-global";
 				reg = <0x981000 0x1000>;
 				status = "disabled";
+				power-domains = <&gpd NRF_GPD_SLOW_MAIN>;
```
Shouldn't DPPIC131 be in `NRF_GPD_SLOW_ACTIVE`, like SAADC, [LP]COMP, TEMP, etc.?

gmarull replied:
yes, good catch

gmarull commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -171,7 +191,21 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+		uint32_t retain;
+
+		retain = nrf_gpio_port_retain_get(cfg->port) | BIT(pin);
+		nrf_gpio_port_retain_set(cfg->port, retain);
```
done

gmarull commented on include/zephyr/dt-bindings/power/nordic-nrf-gpd.h:
```diff
@@ -0,0 +1,16 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_POWER_NORDIC_NRF_GLOBAL_PD
+#define ZEPHYR_INCLUDE_DT_BINDINGS_POWER_NORDIC_NRF_GLOBAL_PD
+
+#define NRF_GPD_SLOW_MAIN 2U
```
added comment

hubertmis commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -95,6 +103,15 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+		ret = nrf_gpd_request(NRF_GPD_FAST_ACTIVE1);
+		if (ret < 0) {
+			return ret;
+		}
+	}
```
It seems we are still missing something like:

```
cfg->port->RETAINCLR = BIT(pin);
```

around here

hubertmis reviewed (changes requested):
The GPIO driver must clear GPIO RETAIN. Otherwise, it will work only for the first request.

Potentially we can figure out GPIO RETAIN in pinctrl out of this PR, but still in 2.8.0 timeline.

masz-nordic commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -171,7 +191,18 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+end:
+#ifdef CONFIG_SOC_NRF54H20_GPD
+	if (cfg->pad_pd == NRF_GPD_FAST_ACTIVE1) {
+		cfg->port->RETAINSET = BIT(pin);
```
Please make sure to use HAL (add new function if needed).

gmarull replied:
done

gmarull committed:
soc: nordic: nrf54h: gpd: add API to set/clear pin retention

This API needs to be called by FAST peripherals before/after
disabling/enabling them.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: serial: nrfx_uarte: set/clear pins retention

When GPD is managed by pinctrl, pins retention needs to be controlled by
the driver to avoid glitches.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

hubertmis commented on soc/nordic/nrf54h/gpd/gpd.c:
```diff
@@ -211,6 +212,33 @@ int nrf_gpd_release(uint8_t id)
 	return onoff_release(&gpd_mgr->mgr);
 }
 
+int nrf_gpd_retain_pins_set(const struct pinctrl_dev_config *pcfg, bool retain)
```
Not sure why this function is in this unit. The function is about configuring pin properties, while the file is about global power domains. How about moving the function to pinctrl?

gmarull committed:
soc: nordic: nrf54h: disable PM_DEVICE_POWER_DOMAIN

It is enabled by default if we enable device PM, but we do not want
this, otherwise we get linker errors (PM subsys, fun guaranteed!).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

karstenkoenig reviewed (commented):
Looks really good, just unsure on the regular output gpio usage

hubertmis commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -171,7 +193,20 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+#ifdef CONFIG_SOC_NRF54H20_GPD
+	if (cfg->pad_pd == NRF_GPD_FAST_ACTIVE1) {
+		if (flags & GPIO_OUTPUT) {
+			nrf_gpio_port_retain_set(cfg->port, BIT(pin));
```
Right, the same procedure (requesting power, clearing retain, configuring pin, setting retain, releasing power) must be implemented for all output configuration functions:
```
	.port_set_masked_raw = gpio_nrfx_port_set_masked_raw,
	.port_set_bits_raw = gpio_nrfx_port_set_bits_raw,
	.port_clear_bits_raw = gpio_nrfx_port_clear_bits_raw,
	.port_toggle_bits = gpio_nrfx_port_toggle_bits,
```

gmarull committed:
modules: hal_nordic: bump regtool to 8.0.0

It is required for some pinctrl changes.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: checkpatch: add NRF_GPIO_Type to typedefs file

So that we do not get false warnings about consistent spacing around
'*'.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

karstenkoenig reviewed (dismissed):
Just minor comments left, looks surprisingly good given the complexity

anangl commented on soc/nordic/nrf54h/gpd/gpd.c:
```diff
@@ -0,0 +1,273 @@
+	nrfs_err_t err;
+	int ret;
+
+	ret = nrfs_backend_wait_for_connection(K_NO_WAIT);
+	if (ret < 0) {
+		return ret;
```
I think that `atomic_set_bit(&gpd_service_status, GPD_SERVICE_ERROR);` is missing here.

anangl commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -352,6 +355,19 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+					return ret;
+				}
+
+				nrf_gpio_pin_retain_clear(pins[i]);
```
```suggestion
				nrf_gpio_pin_retain_clear(psel);
```

anangl commented on include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:
```diff
@@ -10,7 +10,9 @@
- * - 31..18: Pin function.
+ * - 31..24: Pin function.
+ * - 20-23:  Reserved.
+ * - 19:     Associated peripheral belongs to GD FAST ACTIVE1 (nRF54H only)
  * - 17:     Clockpin enable.
```
18 is also reserved?

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -171,7 +234,8 @@ static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
 
-	return 0;
+end:
+	return gpio_nrfx_retain_set(port, BIT(pin), flags);
```
```suggestion
	(void) gpio_nrfx_retain_set(port, BIT(pin), flags);
	return ret;
```

gmarull commented:
last push: dropped HAL changes, will be reintroduced in a follow-up PR

anangl commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -352,6 +355,20 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+#ifdef CONFIG_SOC_NRF54H20_GPD
+			if (NRF_GET_GPD_FAST_ACTIVE1(pins[i]) == 1U) {
+				int ret;
+				NRF_GPIO_Type *port = nrf_gpio_pin_port_decode(&pin);
```
This changes `pin` to a relative number within the `port`, but further `nrf_gpio_*` calls require `pin` in its absolute form. You need to use a local variable for this conversion.

gmarull replied:
thanks good catch

anangl commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -352,6 +355,21 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+					return ret;
+				}
+
+				port->RETAINCLR = BIT(pin);
```
```suggestion
				port->RETAINCLR = BIT(d_pin);
```

hubertmis commented on drivers/gpio/gpio_nrfx.c:
I'm trying to figure out the power domain management required for the input pins.

It seems that `gpio_nrfx_pin_interrupt_configure()` should be enhanced:

If the
```
if (!(BIT(pin) & cfg->edge_sense) &&
	    (mode == GPIO_INT_MODE_EDGE) &&
	    (nrf_gpio_pin_dir_get(abs_pin) == NRF_GPIO_PIN_DIR_INPUT))
```
is met and the driver uses event IN, the relevant power domain should be kept powered up (`nrf_gpd_request()`). Alternatively, if the driver uses event SENSE, the GPIO.RETAIN should be applied to the pin (`cfg->port->RETAINSET = mask`).

However, I'm not sure in which procedure this configuration should be cleared. @nordic-krch, could you help to identify the correct place?

hubertmis reviewed (approved):
The GPIO IN is most probably not handled yet. But I think we can add it in a next PR.

karstenkoenig reviewed (approved):
This wouldn't work for pins defined as input on the fast ports, but that was broken before so probably easier to sort in follow up PR as suggest by Hubert to avoid complicating this any further
